src/kaleidoscope/hid.cpp

```diff
@@ -51,13 +51,86 @@
 namespace kaleidoscope {
 namespace hid {
 
+// Anonymous namespace for internal variables and helper functions
+namespace {
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `mod_flags_allowed`, we can prohibit any modifier flags that aren't in
+// the key most recently pressed, fixing the rollover problem, and getting the intended
+// `The` instead of `THe`. This was particularly noticeable to people who were using a
+// separate layer for `shift`, and a OneShot layer key to activate it, causing what looked
+// like a OneShot bug.
+
+// A bitfield of the current modifier flags from keys processed so far in the current
+// keyswitch scan:
+static byte current_mod_flags{0};
+
+// The current set of modifier flags that will be allowed to be added to the upcoming
+// keyboard HID report. This should be set by the latest key pressed:
+static byte mod_flags_allowed{0};
+
+// The keycode of the last key to toggle on. This get
+static byte toggled_on_keycode{0};
+
+// This returns true if the key is a keyboard key, and its keycode is a modifier.
+bool isPureModifier(Key mappedKey) {
+  // If it's not a keyboard key, return false
+  if (mappedKey.flags & (SYNTHETIC | RESERVED))
+    return false;
+  // An efficient way to check if the keycode is for a modifier. Key_LeftControl has the
+  // modifier keycode with the lowest value, and the others are sequential, and we use an
+  // unsigned integer so after subtraction, the comparison is simple.
+  byte offset_keycode = mappedKey.keyCode - Key_LeftControl.keyCode;
+  return (offset_keycode < 8);
+}
+
+// This function adds modifier flags to the bitfield for current modifier flags that will
+// be added to the report after all keys are scanned.
+void addModFlags(byte flags) {
+  current_mod_flags |= flags;
+}
+
+// This function actually adds the modifier flags to the upcoming report.
+void pressModFlags(byte flags) {
+  if (flags & SHIFT_HELD) {
+    pressRawKey(Key_LeftShift);
+  }
+  if (flags & CTRL_HELD) {
+    pressRawKey(Key_LeftControl);
+  }
+  if (flags & LALT_HELD) {
+    pressRawKey(Key_LeftAlt);
+  }
+  if (flags & RALT_HELD) {
+    pressRawKey(Key_RightAlt);
+  }
+  if (flags & GUI_HELD) {
+    pressRawKey(Key_LeftGui);
+  }
+}
+
+} // anonymous namespace
+
+
 void initializeKeyboard() {
   Keyboard.begin();
   WITH_BOOTKEYBOARD {
     BootKeyboard.begin();
   }
 }
 
+// Called from handleKeyswitchEventDefault when a key toggles on. This sets the mask of
+// modifier flags that are allowed to be added to the upcoming report to the ones in the
+// key that just toggled on, so when we roll over from a key with a modifier flag to one
+// without it, that modifier flag won't affect the new keypress.
+void setLastKeyPress(Key mappedKey) {
```

Comment:
There's something about 'setLastKeypress' that feels wrong, but I can't quite put my finger on it.

Does this do the right thing when two keys are toggled on in the same keyscan cycle? (I'd not believe that was possible if I hadn't just solved another bug that happens in that case).

Reply:
>  Kaleidoglyph doesn't have this problem

Are you able to explain how you solved this in Kaleidoglyph or point me at the relevant commits or bits of code? I'd love to understand what you're doing differently